docs/designdoc/safe_statefulset_rolling_update.md

```diff
@@ -0,0 +1,171 @@
+# Safe Rolling Update of StatefulSet
+
+## Context
+
+In the current specification of MOCO, the following problems have occurred in the past.
+
+* Due to a node failure, one MySQL Pod stopped working, and the MySQLCluster became Unhealthy.
+* In this state, the MySQLCluster was updated, and the StatefulSet's rolling update was executed.
+* Due to the failure and update, two MySQL Pods went down at the same time, and the MySQLCluster became Unavailable.
+* At this time, PDB was set. However, because the StatefulSet update does not consider the PDB or the state of the existing Pods, the MySQL Pod was updated even though the cluster was Unhealthy.
+
+To prevent such problems, MOCO will implement a function to safely update the StatefulSet.
+
+## Goals
+
+* Update Pods considering the state of MySQLCluster when updating StatefulSet
+* Do not involve destructive changes
+
+## ActualDesign
+
+When updating the StatefulSet of MySQLCluster, set `.spec.updateStrategy.rollingUpdate.partition` of StatefulSet dynamically and update Pods considering the state of MySQLCluster.
+
+Partition is a parameter to divide the update of StatefulSet's Pods.
+If you specify `.partition`, updates will be made only for StatefulSet Pods with a number equal to or greater than the specified number.
+If `.partition` is larger than `.spec.replicas`, updates to `.spec.template` will not propagate to that Pod.
+By using this function, you can roll out the update of StatefulSet's Pods in stages.
+
+```yaml
+updateStrategy:
+  type: RollingUpdate
+  rollingUpdate:
+    partition: 3
+```
+
+In MOCO, `updateStrategy` of the StatefulSet of the created MySQLCluster is fixed to `type: RollingUpdate`.
+This document proposes to add a reconciler that dynamically sets `.partition` to realize safe updating of StatefulSet.
+
+### When creating MySQLCluster
+
+Until MySQLCluster successfully starts for the first time, applies StatefulSet without setting `.partition`.
+This is done to prevent a situation where a user-corrected StatefulSet, due to misconfigurations like image issues, does not roll out due to partition.
```

Comment:
Please change this description. Now we set the partition when creating.

Reply:
Thanks, I fixed: https://github.com/cybozu-go/moco/pull/628/commits/111de9712f49d0dec860aa1d0aaaf79f54715257